core/src/main/java/org/opensearch/remote/metadata/client/AbstractSdkClient.java

```diff
@@ -66,4 +101,150 @@ protected boolean shouldUseId(String id) {
+                // return if it's user's resource
+                return CompletableFuture.completedFuture(response);
+            }
+            return addToGlobalResourceCache(request, dataFetched);
+        });
+    }
+
+    /**
+     * Add the global resource to cache.
+     * @param request origin request.
+     * @param dataFetchedWithGlobalTenantId data fetched with global tenant id, which means this resource is a global resource.
+     * @return {@link CompletionStage<GetDataObjectResponse>} The response been added to cache and then replaced the tenant_id to request's tenant id.
+     */
+    protected CompletionStage<GetDataObjectResponse> addToGlobalResourceCache(
+        GetDataObjectRequest request,
+        CompletionStage<GetDataObjectResponse> dataFetchedWithGlobalTenantId
+    ) {
+        return dataFetchedWithGlobalTenantId.thenCompose(res -> {
+            if (globalResourceCacheTTL.getMillis() == 0) {
+                return CompletableFuture.completedFuture(replaceGlobalTenantId(request, res));
+            }
+            GLOBAL_RESOURCES_CACHE.put(buildGlobalCacheKey(request.index(), request.id()), new Tuple<>(res, System.currentTimeMillis()));
+            return CompletableFuture.completedFuture(getGlobalResourceDataFromCache(request));
+        });
+    }
+
+    /**
+     * Read global resource from cache.
+     * @param request origin get request.
+     * @return {@link GetDataObjectResponse} The response cached in cache and replaced with request tenant id.
+     */
+    protected GetDataObjectResponse getGlobalResourceDataFromCache(GetDataObjectRequest request) {
+        String checkingKey = buildGlobalCacheKey(request.index(), request.id());
+        long currentTime = System.currentTimeMillis();
+        Tuple<GetDataObjectResponse, Long> tuple = GLOBAL_RESOURCES_CACHE.computeIfPresent(checkingKey, (key, value) -> {
+            if (currentTime - value.v2() > globalResourceCacheTTL.getMillis()) {
+                return null; // If expired, return null to remove the entry
+            }
+            return value; // Keep the entry
+        });
+        // If we have a value, replace the tenant id in the global resource response to actual tenant id
+        // to bypass the validation of the resources e.g.:
+        // https://github.com/opensearch-project/ml-commons/blob/main/ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLAgentExecutor.java#L206
+        return tuple != null ? replaceGlobalTenantId(request, tuple.v1()) : null;
+    }
+
+    private GetDataObjectResponse replaceGlobalTenantId(GetDataObjectRequest request, GetDataObjectResponse response) {
+        response.source().put(TENANT_ID_FIELD_KEY, request.tenantId());
+        GetResponse getResponse = response.getResponse();
+        if (getResponse == null) {
+            throw new OpenSearchStatusException(
+                "Cached response is null, please check configuration with system admin!",
```

Comment:
I think this message will be super confusing for customers. Can we add some meaningful error message? They don't have any context of cache.